lib/node_modules/@stdlib/repl/lib/tokenizer.js

```diff
@@ -0,0 +1,324 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var parse = require( 'acorn-loose' ).parse;
+var walk = require( 'acorn-walk' );
+var linkedList = require( '@stdlib/utils/linked-list' );
+var contains = require('@stdlib/array/base/assert/contains');
+var resolveLocalScopes = require( './resolve_local_scopes.js' );
+var resolveLocalScope = require( './resolve_local_scope.js' );
+var commands = require( './commands.js' );
+
+
+// VARIABLES //
+
+var COMMANDS = commands();
+var RE_PUNCTUATION = /(\{|\}|\[|\]|\(|\)|,|;|:|\.|\?|\?\.|=>|\.\.\.|`|\${)/;
+
+
+// MAIN //
+
+/**
+* Tokenizes the input line based on ECMAScript specification.
+*
+* @private
+* @name tokenizer
+* @type {Function}
+* @param {string} line - input line
+* @param {Object} context - REPL context
+* @returns {Array} array of tokens
+*/
+function tokenizer( line, context ) {
+	var declarations;
+	var VISITORS;
+	var tokens = [];
+	var ast;
+	var i;
+
+	// Parse the given line into tokens & comments...
+	ast = parse( line, {
+		'ecmaVersion': 'latest',
+		'onToken': onToken,
+		'onComment': onComment
+	});
+
+	// Resolve variable declarations from the given line as tokens...
+	declarations = [];
+	ast = resolveLocalScopes( ast, declarations );
+	if ( declarations ) {
+		for ( i = 0; i < declarations.length; i++ ) {
+			if ( declarations[ i ].start === declarations[ i ].end ) {
+				continue;
+			}
+			tokens.push({
+				'type': 'name',
+				'value': declarations[ i ].name,
+				'start': declarations[ i ].start,
+				'end': declarations[ i ].end
+			});
+		}
+	}
+
+	// Resolve identifiers from the given line as tokens...
+	VISITORS = {
+		'Identifier': Identifier,
+		'MemberExpression': MemberExpression
+	};
+	walk.simple( ast, VISITORS );
+
+	return tokens;
+
+	/**
+	* Callback invoked upon encountering a `Token` when parsing.
+	*
+	* @private
+	* @param {Object} token - token object
+	*/
+	function onToken( token ) {
+		// Ignore non-existent tokens:
+		if ( token.start === token.end ) {
+			return;
+		}
+		if ( token.type.isLoop || contains( [ 'if', 'else', 'switch', 'case', 'catch', 'finally', 'try', 'return', 'break', 'continue' ], token.type.keyword ) ) {
```

Comment:
This list is a fixed list. Use `contains.factory` to perform partial application and improve perf.